File: tls-certificate-expiring/node/check-tls-expiry.js
/**
 * Check a host's live TLS certificate expiry over a raw socket, and if the
 * failure is a missing or wrong CAA record, repair it through the Cloudflare
 * DNS API. Safe by default: DRY_RUN only reports the plan until turned off.
 */
import { pathToFileURL } from "node:url";

const DNS_DOMAIN = process.env.DNS_DOMAIN || "example.com";
const CLOUDFLARE_API_TOKEN = process.env.CLOUDFLARE_API_TOKEN || "";
const CLOUDFLARE_ZONE_ID = process.env.CLOUDFLARE_ZONE_ID || "";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const CA_TO_PERMIT = process.env.CA_TO_PERMIT || "letsencrypt.org";
const WARN_AT_DAYS = Number(process.env.WARN_AT_DAYS || 21);
const CRIT_AT_DAYS = Number(process.env.CRIT_AT_DAYS || 7);

/** Pure function, no I/O. Both dates should be Date objects. */
export function daysUntilExpiry(notAfter, now) {
  const msPerDay = 24 * 60 * 60 * 1000;
  return Math.floor((notAfter.getTime() - now.getTime()) / msPerDay);
}

/** Pure function, no I/O. Classifies remaining days into a severity. */
export function classify(days, warnAt = 21, critAt = 7) {
  if (days < 0) return "expired";
  if (days <= critAt) return "critical";
  if (days <= warnAt) return "warn";
  return "ok";
}

/** Open a raw TLS socket with SNI set and read the peer certificate. */
async function fetchPeerCertificate(host, port = 443) {
  const tls = await import("node:tls");
  return new Promise((resolve, reject) => {
    const socket = tls.connect(
      { host, port, servername: host, timeout: 15000 },
      () => {
        const cert = socket.getPeerCertificate();
        socket.end();
        resolve(cert);
      }
    );
    socket.on("error", reject);
    socket.on("timeout", () => {
      socket.destroy();
      reject(new Error("TLS connection timed out"));
    });
  });
}

/** Return the list of CA domains permitted to issue for this name via CAA. */
async function checkCaa(domain) {
  const dns = await import("node:dns/promises");
  try {
    const records = await dns.resolveCaa(domain);
    return records.map((r) => r.issue).filter(Boolean);
  } catch (err) {
    if (err.code === "ENODATA" || err.code === "ENOTFOUND") return [];
    throw err;
  }
}

/** Add a CAA record permitting caDomain to issue for domain, via Cloudflare. */
async function addCaaRecord(domain, caDomain) {
  if (DRY_RUN) {
    console.log(`[dry run] would add CAA record on ${domain} permitting ${caDomain}`);
    return;
  }

  const url = `https://api.cloudflare.com/client/v4/zones/${CLOUDFLARE_ZONE_ID}/dns_records`;
  const res = await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${CLOUDFLARE_API_TOKEN}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      type: "CAA",
      name: domain,
      data: { flags: 0, tag: "issue", value: caDomain },
    }),
  });
  if (!res.ok) throw new Error(`Cloudflare API returned ${res.status}`);
  console.log(`Added CAA record on ${domain} permitting ${caDomain}`);
}

export async function run() {
  console.log(`Checking TLS certificate for ${DNS_DOMAIN} (DRY_RUN=${DRY_RUN})`);

  const cert = await fetchPeerCertificate(DNS_DOMAIN);
  const notAfter = new Date(cert.valid_to);
  const now = new Date();
  const days = daysUntilExpiry(notAfter, now);
  const severity = classify(days, WARN_AT_DAYS, CRIT_AT_DAYS);

  console.log(`Certificate for ${DNS_DOMAIN} expires in ${days} day(s): ${severity}`);

  if (severity === "ok") {
    console.log("OK: certificate has plenty of runway left.");
    return;
  }

  console.warn(`Certificate for ${DNS_DOMAIN} is ${severity} (${days} days remaining).`);

  const permittedCas = await checkCaa(DNS_DOMAIN);
  if (permittedCas.length > 0 && !permittedCas.some((ca) => ca.includes(CA_TO_PERMIT))) {
    console.warn(
      `CAA record on ${DNS_DOMAIN} only permits ${JSON.stringify(permittedCas)}, which blocks issuance from ${CA_TO_PERMIT}.`
    );
    if (CLOUDFLARE_API_TOKEN && CLOUDFLARE_ZONE_ID) {
      await addCaaRecord(DNS_DOMAIN, CA_TO_PERMIT);
    } else {
      console.warn("Set CLOUDFLARE_API_TOKEN and CLOUDFLARE_ZONE_ID to let this script add the CAA record.");
    }
  } else {
    console.warn(
      "CAA looks fine. Check the ACME client and its port 80/443 or DNS-01 automation by hand: " +
      "run 'sudo certbot renew --dry-run' on the host to see the exact failure."
    );
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((e) => { console.error(e); process.exit(1); });
}

File: tls-certificate-expiring/python/check_tls_expiry.py
"""Check a host's live TLS certificate expiry over a raw socket, and if the
failure is a missing or wrong CAA record, repair it through the Cloudflare
DNS API. Safe by default: DRY_RUN only reports the plan until turned off.
"""
import os
import logging
from datetime import datetime, timezone
from typing import Literal

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("check_tls_expiry")

DNS_DOMAIN = os.environ.get("DNS_DOMAIN", "example.com")
CLOUDFLARE_API_TOKEN = os.environ.get("CLOUDFLARE_API_TOKEN", "")
CLOUDFLARE_ZONE_ID = os.environ.get("CLOUDFLARE_ZONE_ID", "")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
CA_TO_PERMIT = os.environ.get("CA_TO_PERMIT", "letsencrypt.org")
WARN_AT_DAYS = int(os.environ.get("WARN_AT_DAYS", "21"))
CRIT_AT_DAYS = int(os.environ.get("CRIT_AT_DAYS", "7"))


def days_until_expiry(not_after: datetime, now: datetime) -> int:
    """Pure function, no I/O. Both datetimes should be timezone-aware."""
    return (not_after - now).days


def classify(
    days: int, warn_at: int = 21, crit_at: int = 7
) -> Literal["ok", "warn", "critical", "expired"]:
    """Pure function, no I/O. Classifies remaining days into a severity."""
    if days < 0:
        return "expired"
    if days <= crit_at:
        return "critical"
    if days <= warn_at:
        return "warn"
    return "ok"


def fetch_peer_certificate(host: str, port: int = 443):
    """Open a raw TLS socket with SNI set and read the peer certificate."""
    import socket
    import ssl

    context = ssl.create_default_context()
    with socket.create_connection((host, port), timeout=15) as sock:
        with context.wrap_socket(sock, server_hostname=host) as tls_sock:
            return tls_sock.getpeercert()


def parse_not_after(cert: dict) -> datetime:
    """Parse the certificate's notAfter field (e.g. 'Sep  1 00:00:00 2026 GMT')."""
    from datetime import datetime as dt

    return dt.strptime(cert["notAfter"], "%b %d %H:%M:%S %Y %Z").replace(tzinfo=timezone.utc)


def check_caa(domain: str) -> list:
    """Return the list of CA domains permitted to issue for this name via CAA."""
    import dns.resolver

    try:
        answers = dns.resolver.resolve(domain, "CAA")
    except dns.resolver.NoAnswer:
        return []
    except dns.resolver.NXDOMAIN:
        return []
    permitted = []
    for rdata in answers:
        tag = rdata.tag.decode() if isinstance(rdata.tag, bytes) else rdata.tag
        if tag == "issue":
            value = rdata.value.decode() if isinstance(rdata.value, bytes) else rdata.value
            permitted.append(value)
    return permitted


def add_caa_record(domain: str, ca_domain: str) -> None:
    """Add a CAA record permitting ca_domain to issue for domain, via Cloudflare."""
    import requests

    if DRY_RUN:
        log.info("[dry run] would add CAA record on %s permitting %s", domain, ca_domain)
        return

    url = f"https://api.cloudflare.com/client/v4/zones/{CLOUDFLARE_ZONE_ID}/dns_records"
    headers = {
        "Authorization": f"Bearer {CLOUDFLARE_API_TOKEN}",
        "Content-Type": "application/json",
    }
    payload = {
        "type": "CAA",
        "name": domain,
        "data": {"flags": 0, "tag": "issue", "value": ca_domain},
    }
    r = requests.post(url, headers=headers, json=payload, timeout=30)
    r.raise_for_status()
    log.info("Added CAA record on %s permitting %s", domain, ca_domain)


def run():
    log.info("Checking TLS certificate for %s (DRY_RUN=%s)", DNS_DOMAIN, DRY_RUN)

    cert = fetch_peer_certificate(DNS_DOMAIN)
    not_after = parse_not_after(cert)
    now = datetime.now(timezone.utc)
    days = days_until_expiry(not_after, now)
    severity = classify(days, WARN_AT_DAYS, CRIT_AT_DAYS)

    log.info("Certificate for %s expires in %d day(s): %s", DNS_DOMAIN, days, severity)

    if severity == "ok":
        log.info("OK: certificate has plenty of runway left.")
        return

    log.warning("Certificate for %s is %s (%d days remaining).", DNS_DOMAIN, severity, days)

    permitted_cas = check_caa(DNS_DOMAIN)
    if permitted_cas and not any(CA_TO_PERMIT in ca for ca in permitted_cas):
        log.warning(
            "CAA record on %s only permits %s, which blocks issuance from %s.",
            DNS_DOMAIN, permitted_cas, CA_TO_PERMIT,
        )
        if CLOUDFLARE_API_TOKEN and CLOUDFLARE_ZONE_ID:
            add_caa_record(DNS_DOMAIN, CA_TO_PERMIT)
        else:
            log.warning(
                "Set CLOUDFLARE_API_TOKEN and CLOUDFLARE_ZONE_ID to let this script add the CAA record."
            )
    else:
        log.warning(
            "CAA looks fine. Check the ACME client and its port 80/443 or DNS-01 automation by hand: "
            "run 'sudo certbot renew --dry-run' on the host to see the exact failure."
        )


if __name__ == "__main__":
    run()
